crates/compute/src/layer.rs

```diff
@@ -122,6 +160,77 @@ pub trait ComputeLayer<F> {
 		coordinates: &[F],
 		data: &mut <Self::DevMem as ComputeMemory<F>>::FSliceMut<'_>,
 	) -> Result<(), Error>;
+
+	// TODO: better docs
+	fn sum_composition_evals(
+		&self,
+		exec: &mut Self::KernelExec,
+		log_len: usize,
+		inputs: &[KernelBuffer<'_, F, Self::DevMem>],
```

Comment:
In the original version, this had the type `&[Self::FSlice<'_>]`. It seems this is the more appropriate type

Reply:
I think we should stick with `Self::FSlice`. We can always cast an `FSliceMut` to an `FSlice` with `ComputeMemory::as_const`. Maybe we should even have a method that helps convert a `KernelBuffer` to an `FSlice` (passing through Ref and calling `as_const` on Mut).

Reply:
I reverted to `FSlice` and moved conversion to the test fn.

Additionally, I've added fns to `KernelBuffer` to extract `Ref` and `Mut` variants. Note that it's not quite possible (at least I don't see it right now) to directly convert `KernelBuffer::Mut` to `FSlice`: if we fant to return `FSlice` (in `Ref` case) we need to take ownership of `KernelBuffer`, but `Mem::as_const` wants a reference, so we cannot have a fn that does both conversions...

I did not fully understand the bit about casting `FSliceMut` to `FSlice` wrt. to this discussion.

Reply:
Pushed a commit to handle both cases.